### PR TITLE
Adds Superposition task 1.14

### PR DESCRIPTION
**Pull request by frtibble**
Part of #1185

### CODE REVIEW

File: katas/content/index.json [hunk 1]
```diff
@@ -12,5 +12,6 @@
   "oracles",
   "deutsch_algo",
   "deutsch_jozsa",
-  "qec_shor"
+  "qec_shor",
+  "superposition"
```
Comment (frtibble) [inline katas/content/index.json:16]: TODO: remove before merge

File: katas/content/superposition/parity_bitstrings/solution.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
+We are going to use a recursive approach that was hinted at in the task.
+
+Let's denote the required state on $N$ qubits as $|P_{N,0}\rangle$ for `parity = 0` and $|P_{N,1}\rangle$ for `parity = 1`. We can group the $2^{N-1}$ basis states included in the state $|P_{N,0}\rangle$ by their first bit ($0$ or $1$) and write the state as follows:
+
+$$|P_{N,0}\rangle = \frac{1}{\sqrt{2^{N-1}}} \sum_{k : k \text{ has parity }0} |k\rangle_N = \\\\
```
Comment (tcNickolas) [inline katas/content/superposition/parity_bitstrings/solution.md:5]: \\\\ don't work in plain formulas right now (only in special environments like matrix or cases), so let's make each line a separate formula instead

File: katas/content/superposition/parity_bitstrings/solution.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
+We are going to use a recursive approach that was hinted at in the task.
```
Comment (tcNickolas) [inline katas/content/superposition/parity_bitstrings/solution.md:1]:
```suggestion
There are multiple ways to approach this problem. In our first solution, we are going to use a recursive approach that was hinted at in the task.
```

File: katas/content/superposition/parity_bitstrings/solution.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+$$|P_{N,0}\rangle = \frac{1}{\sqrt{2^{N-1}}} \sum_{k : k \text{ has parity }0} |k\rangle_N = \\\\
+= \frac{1}{\sqrt{2^{N-1}}} \big( |0\rangle \otimes \sum_{k' : k' \\text{ has parity }0} |k'\rangle_{N-1} + |1\rangle \otimes \sum_{k'' : k'' \\text{ has parity }1} |k''\rangle_{N-1} \big) = \\\\
+= \frac{1}{\sqrt{2}} \big( |0\rangle \otimes |P_{N-1,0}\rangle + |1\rangle \otimes |P_{N-1,1}\rangle \big)$$
+We can consider the expression for $|P_{N,1}\rangle$ in a similar manner, and get a unified expression for both states:
+$$|P_{N,p}\rangle = \frac{1}{\sqrt{2}} \big( |0\rangle \otimes |P_{N-1,p}\rangle + |1\rangle \otimes |P_{N-1,1-p}\rangle \big)$$
+
+Now we can use this expression to prepare the state using a recursive approach we've seen before:
+
+1. Apply **H** gate to the first qubit to prepare state $\frac{1}{\sqrt{2}} ( |0\rangle + |1\rangle ) \otimes |0\rangle_{N-1}$
```
Comment (tcNickolas) [inline katas/content/superposition/parity_bitstrings/solution.md:13]:
```suggestion
1. Apply $H$ gate to the first qubit to prepare state $\frac{1}{\sqrt{2}} ( |0\rangle + |1\rangle ) \otimes |0\rangle_{N-1}$.
```

File: katas/content/superposition/parity_bitstrings/solution.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+2. Apply the controlled variant of procedure of preparing $|P_{N-1,p}\rangle$ on the last $N-1$ qubits, with the first qubit in the $|0\rangle$ state as the control.
+3. Apply the controlled variant of procedure of preparing $|P_{N-1,1-p}\rangle$ on the last $N-1$ qubits, with the first qubit in the $|1\rangle$ state as the control.
+
+> Q# library function [`ApplyControlledOnInt`](https://learn.microsoft.com/qsharp/api/qsharp-lang/microsoft.quantum.canon/applycontrolledonint) allows to do that easily.
+
+4. The base of recursion is preparing the states for $N = 1$:
+* For `parity = 0`, there is one single-qubit state with this parity: $|0\rangle$ (no action required to prepare).
+* For `parity = 1`, there is one single-qubit state with this parity: $|1\rangle$ (apply **X** gate to prepare).
```
Comment (tcNickolas) [inline katas/content/superposition/parity_bitstrings/solution.md:21]:
```suggestion
* For `parity = 1`, there is one single-qubit state with this parity: $|1\rangle$ (apply $X$ gate to prepare).
```

File: katas/content/superposition/parity_bitstrings/solution.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+@[solution]({ "id": "superposition__parity_bitstrings_solution_a", "codePath": "./SolutionA.qs" })
+
+We start by preparing an equal superposition of all basis states and allocating an extra qubit.
```
Comment (tcNickolas) [inline katas/content/superposition/parity_bitstrings/solution.md:25]:
```suggestion
In the second solution, we'll use post-selection. We start by preparing an equal superposition of all basis states and allocating an extra qubit.
```

File: katas/content/superposition/parity_bitstrings/solution.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+@[solution]({ "id": "superposition__parity_bitstrings_solution_b", "codePath": "./SolutionB.qs" })
+
+Yet another way of getting the desired superposition could be preparing the mix of all possible basis states for all qubits iteratively, keeping the parity on each step.
+
+We start by preparing a superposition of all basis states with parity $0$. To achieve that, we loop through all the qubits except the first one and prepare all of them in equal superposition. In order to maintain the parity of the basis states involved, we use the first qubit and conditionally flip its state using a CNOT gate with each next qubit as the control, so that every time there is a $|1\rangle$ state in the chain, we get back to an even number of $|1\rangle$ states in that basis state.
+
+For example, after the first loop iteration we get the state $\frac12(|00\rangle + |11\rangle)$. After the second iteration we get the state
+
+$$CNOT_{2,0} \frac1{\sqrt2}(|00\rangle + |11\rangle) \otimes \frac1{\sqrt2}(|0\rangle + |1\rangle) = \\\\
```
Comment (tcNickolas) [inline katas/content/superposition/parity_bitstrings/solution.md:41]: We're doing to stop using \color (see https://github.com/microsoft/qsharp/issues/1435), so let's use \textbf to highlight the parts we want to highlight instead (first and third qubits I believe? color formatting doesn't really work when rendering in the playground or in VS Code notebooks)

**Review by tcNickolas — CHANGES_REQUESTED** [review]
Looks good overall, I left some formatting and testing comments.

Thank you!

File: katas/content/superposition/parity_bitstrings/Verification.qs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        } else {
+            H(qs[0]);
+            ApplyControlledOnInt(0, AllStatesWithParitySuperposition_Reference, qs[0 .. 0], (qs[1 ...], parity));
+            ApplyControlledOnInt(1, AllStatesWithParitySuperposition_Reference, qs[0 .. 0], (qs[1 ...], 1 - parity));
+        }
+    }
+
+    @EntryPoint()
+    operation CheckSolution() : Bool {
+        Message($"Testing for N = 2...");
```
[the remark this answers is not in the document]
Comment (frtibble) [reply katas/content/superposition/parity_bitstrings/Verification.qs:19]: Sounds good - I've moved the message inside the parity loop, added parity in the message, removed the i... loop, and combined the cases for N

File: katas/content/superposition/parity_bitstrings/solution.md (hunk 7 above)
Comment (frtibble) [reply katas/content/superposition/parity_bitstrings/solution.md:41]: Noted!

File: katas/content/superposition/parity_bitstrings/solution.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
+There are multiple ways to approach this problem. In our first solution, we are going to use a recursive approach that was hinted at in the task.
+
+Let's denote the required state on $N$ qubits as $|P_{N,0}\rangle$ for `parity = 0` and $|P_{N,1}\rangle$ for `parity = 1`. We can group the $2^{N-1}$ basis states included in the state $|P_{N,0}\rangle$ by their first bit ($0$ or $1$) and write the state as follows:
+
+$$|P_{N,0}\rangle = \frac{1}{\sqrt{2^{N-1}}} \sum_{k : k \text{ has parity }0} |k\rangle_N = $$
+$$= \frac{1}{\sqrt{2^{N-1}}} \big( |0\rangle \otimes \sum_{k' : k' \\text{ has parity }0} |k'\rangle_{N-1} + |1\rangle \otimes \sum_{k'' : k'' \\text{ has parity }1} |k''\rangle_{N-1} \big) = $$
```
Comment (tcNickolas) [inline katas/content/superposition/parity_bitstrings/solution.md:6]: These won't affect the current parser but the change in an adjacent pull request starts treating these as new lines, so let's preempt that and use single-backslashes:
```suggestion
$$= \frac{1}{\sqrt{2^{N-1}}} \big( |0\rangle \otimes \sum_{k' : k' \text{ has parity }0} |k'\rangle_{N-1} + |1\rangle \otimes \sum_{k'' : k'' \text{ has parity }1} |k''\rangle_{N-1} \big) = $$
```

**Review by tcNickolas — APPROVED** [review]
Looks great, thank you! I'll apply the final suggestions and merge

File: katas/content/index.json (hunk 1 above)
Comment (tcNickolas) [inline katas/content/index.json:16]:
```suggestion
  "qec_shor"
```

**Pull request merged**